File: API/0087_Parabolic_SAR_Reversal/ParabolicSarReversalStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies
{
    /// <summary>
    /// Parabolic SAR Reversal Strategy.
    /// Enters long when SAR switches from above to below price.
    /// Enters short when SAR switches from below to above price.
    /// </summary>
    public class ParabolicSarReversalStrategy : Strategy
    {
        private readonly StrategyParam<decimal> _initialAcceleration;
        private readonly StrategyParam<decimal> _maxAcceleration;
        private readonly StrategyParam<DataType> _candleType;
        
        private bool? _prevIsSarAbovePrice;

        /// <summary>
        /// Initial acceleration factor for Parabolic SAR.
        /// </summary>
        public decimal InitialAcceleration
        {
            get => _initialAcceleration.Value;
            set => _initialAcceleration.Value = value;
        }

        /// <summary>
        /// Maximum acceleration factor for Parabolic SAR.
        /// </summary>
        public decimal MaxAcceleration
        {
            get => _maxAcceleration.Value;
            set => _maxAcceleration.Value = value;
        }

        /// <summary>
        /// Type of candles to use.
        /// </summary>
        public DataType CandleType
        {
            get => _candleType.Value;
            set => _candleType.Value = value;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="ParabolicSarReversalStrategy"/>.
        /// </summary>
        public ParabolicSarReversalStrategy()
        {
            _initialAcceleration = Param(nameof(InitialAcceleration), 0.02m)
                .SetDisplay("Initial Acceleration", "Initial acceleration factor for Parabolic SAR", "SAR Settings")
                .SetRange(0.01m, 0.05m, 0.01m)
                .SetCanOptimize(true);
                
            _maxAcceleration = Param(nameof(MaxAcceleration), 0.2m)
                .SetDisplay("Max Acceleration", "Maximum acceleration factor for Parabolic SAR", "SAR Settings")
                .SetRange(0.1m, 0.3m, 0.05m)
                .SetCanOptimize(true);
                
            _candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(15).TimeFrame())
                .SetDisplay("Candle Type", "Type of candles to use", "General");
        }

        /// <inheritdoc />
        public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
        {
            return [(Security, CandleType)];
        }

        /// <inheritdoc />
        protected override void OnStarted(DateTimeOffset time)
        {
            base.OnStarted(time);

            // Initialize previous state
            _prevIsSarAbovePrice = null;

            // Create Parabolic SAR indicator
            var parabolicSar = new ParabolicSar
            {
                Acceleration = InitialAcceleration,
                MaxAcceleration = MaxAcceleration
            };

            // Create subscription
            var subscription = SubscribeCandles(CandleType);
            
            // Bind indicator and process candles
            subscription
                .Bind(parabolicSar, ProcessCandle)
                .Start();
                
            // Setup chart visualization
            var area = CreateChartArea();
            if (area != null)
            {
                DrawCandles(area, subscription);
                DrawIndicator(area, parabolicSar);
                DrawOwnTrades(area);
            }
        }

        /// <summary>
        /// Process candle with Parabolic SAR value.
        /// </summary>
        /// <param name="candle">Candle.</param>
        /// <param name="sarValue">Parabolic SAR value.</param>
        private void ProcessCandle(ICandleMessage candle, decimal sarValue)
        {
            // Skip unfinished candles
            if (candle.State != CandleStates.Finished)
                return;

            // Check if strategy is ready to trade
            if (!IsFormedAndOnlineAndAllowTrading())
                return;

            // Determine if SAR is above or below price
            bool isSarAbovePrice = sarValue > candle.ClosePrice;
            
            // If this is the first calculation, just store the state
            if (_prevIsSarAbovePrice == null)
            {
                _prevIsSarAbovePrice = isSarAbovePrice;
                return;
            }
            
            // Check for SAR reversal
            bool sarSwitchedBelow = _prevIsSarAbovePrice.Value && !isSarAbovePrice;
            bool sarSwitchedAbove = !_prevIsSarAbovePrice.Value && isSarAbovePrice;
            
            // Long entry: SAR switched from above to below price
            if (sarSwitchedBelow && Position <= 0)
            {
                BuyMarket(Volume + Math.Abs(Position));
                LogInfo($"Long entry: SAR ({sarValue}) switched below price ({candle.ClosePrice})");
            }
            // Short entry: SAR switched from below to above price
            else if (sarSwitchedAbove && Position >= 0)
            {
                SellMarket(Volume + Math.Abs(Position));
                LogInfo($"Short entry: SAR ({sarValue}) switched above price ({candle.ClosePrice})");
            }
            
            // Update the previous state
            _prevIsSarAbovePrice = isSarAbovePrice;
        }
    }
}

File: API/0087_Parabolic_SAR_Reversal/PY/parabolic_sar_reversal_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ParabolicSar
from StockSharp.Algo.Strategies import Strategy

class parabolic_sar_reversal_strategy(Strategy):
    """
    Parabolic SAR Reversal strategy.
    Enters long when SAR switches from above to below price.
    Enters short when SAR switches from below to above price.
    Uses cooldown to control trade frequency.
    """

    def __init__(self):
        super(parabolic_sar_reversal_strategy, self).__init__()
        self._acceleration = self.Param("Acceleration", 0.02).SetDisplay("Acceleration", "Initial acceleration factor", "SAR")
        self._acceleration_max = self.Param("AccelerationMax", 0.2).SetDisplay("Max Acceleration", "Maximum acceleration factor", "SAR")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(1))).SetDisplay("Candle Type", "Type of candles to use", "General")
        self._cooldown_bars = self.Param("CooldownBars", 500).SetDisplay("Cooldown Bars", "Bars to wait between trades", "General")

        self._prev_sar_above = None
        self._cooldown = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(parabolic_sar_reversal_strategy, self).OnReseted()
        self._prev_sar_above = None
        self._cooldown = 0

    def OnStarted2(self, time):
        super(parabolic_sar_reversal_strategy, self).OnStarted2(time)

        self._prev_sar_above = None
        self._cooldown = 0

        sar = ParabolicSar()
        sar.Acceleration = self._acceleration.Value
        sar.AccelerationMax = self._acceleration_max.Value

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(sar, self._process_candle).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, sar)
            self.DrawOwnTrades(area)

    def _process_candle(self, candle, sar_val):
        if candle.State != CandleStates.Finished:
            return

        sv = float(sar_val)
        close = float(candle.ClosePrice)
        is_sar_above = sv > close

        if self._prev_sar_above is None:
            self._prev_sar_above = is_sar_above
            return

        if self._cooldown > 0:
            self._cooldown -= 1
            self._prev_sar_above = is_sar_above
            return

        cd = self._cooldown_bars.Value

        # SAR switched from above to below = bullish signal
        sar_switched_below = self._prev_sar_above == True and not is_sar_above
        # SAR switched from below to above = bearish signal
        sar_switched_above = self._prev_sar_above == False and is_sar_above

        if self.Position == 0 and sar_switched_below:
            self.BuyMarket()
            self._cooldown = cd
        elif self.Position == 0 and sar_switched_above:
            self.SellMarket()
            self._cooldown = cd
        elif self.Position > 0 and sar_switched_above:
            self.SellMarket()
            self._cooldown = cd
        elif self.Position < 0 and sar_switched_below:
            self.BuyMarket()
            self._cooldown = cd

        self._prev_sar_above = is_sar_above

    def CreateClone(self):
        return parabolic_sar_reversal_strategy()
